Feature/2721 show error and output logs

orlandohohmeier opened the pull request:
![ouput-error-log-download](https://cloud.githubusercontent.com/assets/647035/11532077/96ac660c-9901-11e5-9919-e69b1e6d8aae.gif)
Provide links to download logs within the marathon ui, to streamline the task debugging.

**AC**
* [ ] Task/Instance Detail / Get and display list of task files
* [ ] App Detail / Provide error/output links
* [ ] "Cache" mesos calls to improve performance
* [ ] ~~Indicate (file/state) loading progress (**WIP**)~~
* [ ]  Test components, store and helpers (**WIP**)
* [ ]  Changelog entry (**WIP**)

Closes mesosphere/marathon#2721

aldipower commented on src/js/actions/MesosActions.js:
```diff
@@ -0,0 +1,48 @@
+import AppDispatcher from "../AppDispatcher";
```
Could you still use the old require-function to be consistent with our old code.
That has the reason, not to switch between syntax if you work on 2 files with old and new syntax.

I will create a tech debt issue to refactor the require to the new import-statement all at once.

aldipower replied:
https://github.com/mesosphere/marathon/issues/2738

orlandohohmeier replied:
:+1:

aldipower commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+    var agentId = task.slaveId;
+    var taskId = task.id || task.taskId;
+    if (this.state.files == null) {
+      MesosActions.requestTaskFiles(agentId, taskId);
```
Super!

aldipower commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+    var task = this.props.task;
+    var agentId = task.slaveId;
+    var taskId = task.id || task.taskId;
+    if (this.state.files == null) {
```
Sure this check is meaningful? Don't the files path change?

aldipower commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+    var task = this.props.task;
+    var taskId = task.id || task.taskId;
+    var files = MesosStore.getTaskFiles(taskId);
+    if (files != null) {
```
It's called onMesosChange, so there should always be a change, means setState is mandatory.

aldipower commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+
+  getFileNodes() {
+    var files = this.state.files;
+    if (files != null) {
```
files should never be ```null```, you initialize it in getInitialState.

aldipower replied:
Let the store do some magic.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,188 @@
+const STATE_TTL = 60000;
+const MASTER_ID = "master";
+
+class MesosStore extends EventEmitter {
```
We didn't introduced es6-classes yet. Could you write it as the other stores, thanx.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,188 @@
+
+  constructor() {
+    super();
+    this._state = {};
```
The Store is a singleton, so it doesn't make sense to bind this vars at the object-instances.
Please put them in the private scope outside the MesosStore-object.

aldipower replied:
Hmm, it could makes sense, sorry. Depends. But don't use the ```_```, thx.

aldipower commented:
Ok, I'll continue on monday with reviewing, branch is not finished yet.

orlandohohmeier commented:
@aldipower Ok, thanks for your early feedback!

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+  componentDidMount() {
+    var task = this.props.task;
+    var agentId = task.slaveId;
+    var taskId = task.id || task.taskId;
```
Why is this even necessary, again?

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+  getFileNodes() {
+    var files = this.state.files;
+    if (files != null) {
+      return files.map((file) => {
```
2 small things :
- for single arguments, you can skip the brackets
- for direct return statements, you can skip the curlies

So this is preferrable:
```
return files.map(file => (
  <tr key={file.path}>
// ...
));
```

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,86 @@
+    var files = this.state.files;
+    if (files != null) {
+      return files.map((file) => {
+        return (<tr key={file.path}>
```
In any case please break the line when returning JSX:

```
return (
  <tr> // ...
);
```

pierlo-upitup commented on src/js/events/MesosEvents.js:
```diff
@@ -0,0 +1,10 @@
+const MesosEvents = Object.freeze({
+  CHANGE: "MESOS_EVENTS_CHANGE",
+  REQUEST_STATE_COMPLETE: "MESOS_EVENTS_REQUEST_STATE",
```
should be `MESOS_EVENTS_REQUEST_STATE_COMPLETE` for consistency

pierlo-upitup commented on src/js/helpers/JSONPUtil.js:
```diff
@@ -0,0 +1,44 @@
+  /**
+   * Request JSONP data
+   *
+   * @todo: write test to verify this util is working properly
```
Yes please! ;-)

pierlo-upitup replied:
we can probably revisit this once we make progress on https://github.com/mesosphere/marathon/issues/2099

orlandohohmeier replied:
Yep, but as long as we don't have an infrastructure featuring a real DOM, testing will be hard. We could extract some of it's functionality (e.g. building the url) into a util and unit test those methods. What do you think?

pierlo-upitup commented on src/js/helpers/JSONPUtil.js:
```diff
@@ -0,0 +1,42 @@
+      var callback = `jsonp_${Date.now().toString(16)}`;
+      var script = document.createElement("script");
+      // Add jsonp callback
+      window[callback] = function handleJSONPResponce(data) {
```
`Response` please ;)

orlandohohmeier replied:
Ups... :laughing:

pierlo-upitup commented:
Just a question regarding the JSONP module. Wouldn't it make sense to just pull a library that saves us from having to reinvent the wheel and write tests etc.? I've found this one that seems pretty OK https://github.com/bermi/jsonp-client . @orlandohohmeier what do you think?

orlandohohmeier commented:
@pierlo-upitup Yeah, it would make perfect sense to pull a library for the JSONP stuff! Testing the util would be a tough task with our current test infrastructure.
Mesos uses `jsonp` as the callback query param, thus we need a library which is able to handle different callback params. 
The lib you proposed, doesn't handle the callback stuff (registering and handling the jsonp callback) and is not able to handle callback params other than `callback`... But we should have a look and try to find a  suitable lib.  We should also consider to replacing quajax with a lib that is also able to handle JSONP, to have only one library for all the web API calls.

aldipower commented:
@orlandohohmeier 
I don't like the idea of replacing qajax only for the sake of having on library. The usages of qajax is proven now since we introduced this months ago. And it took us a while to find a good ajax-lib besides jQuery. What we could do is to extend the qajax-wrapper (maybe rename it) with the jsonp functionality. The we would achieve one common API for our requests.

orlandohohmeier commented:
@aldipower Yeah, thats a good idea. Any preferences for a new qajax/jsponp-wrapper name?

aldipower commented:
@orlandohohmeier Hmm. :) ```httpRequest``` ? I am not that creative today. ^^ 
I also would suggest to do this in a seperate PR afterwards.

orlandohohmeier commented:
@aldipower Ok, then let's postpone this.

aldipower commented:
Cool. :) This PR is already very large, let's try to keep PRs small, if possible.
Sometimes it's not possible of course..

orlandohohmeier commented:
Yeah, this one is grown into a monster... :japanese_ogre: Sorry about that.

aldipower commented:
Everything fine, this PR is one of the must larger..

pierlo-upitup commented:
@orlandohohmeier I guess we can remove the DO NOT MERGE from the title?

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  getState: function (nodeId) {
+    var data = stateMap[nodeId];
+    if (data == null || data.timestamp + STATE_TTL < Date.now()) {
+      invalidateState(nodeId);
```
Let's keep it simple in this case it's just a null assingment:
```
stateMap[nodeId] = null;
```

orlandohohmeier replied:
The `invalidateState` is used in different places and feels more explicit than simple null assignments.

aldipower replied:
Ah ok, overseen the multiple usage. Sorry.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+
+  getState: function (nodeId) {
+    var data = stateMap[nodeId];
+    if (data == null || data.timestamp + STATE_TTL < Date.now()) {
```
The condition can have this case:
```javascript
if (data == null) {
 data = null;
}
```
Huh?

orlandohohmeier replied:
Yeah, that could happen. Would you recommend to split the condition?

aldipower replied:
Yes, please rewrite it to eliminate the confusion, otherwise I hadn't commented on that.
Thx!

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    return data.state;
+  },
+
+  getTaskFiles: function (taskId) {
```
At lot of repetition compared to getState.
This should be refactored into one function.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  stateMap[nodeId] = null;
+}
+
+function addTaskFiles(taskId, files) {
```
Refactor alongside with ```addState```, too much repetition.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  getTaskFiles: function (taskId) {
+    var data = taskFileMap[taskId];
+    if (data == null || data.timestamp + FILES_TTL < Date.now()) {
+      invalidateTaskFiles(taskId);
```
Please use ```delete``` directly and get rid of the function call. It's only used once.

aldipower replied:
Why are you using 2 ways of invalidating. Comparing this
stateMap[nodeId] = null;
vs.
delete taskFileMap[taskId];
?

orlandohohmeier replied:
The chance is high, that tasks (ids) change over time, therefore it's makes sense to actually remove the task completely from the list. On the other hand, we probably only update the node state and overwriting the reference is more performant in that case.

aldipower replied:
OK, please decide for one method.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+function getNodeUrl(nodeId) {
+  var master = MesosStore.getState(MASTER_ID);
+  if (master == null) {
+    throw new Error(
```
This would break the application!
Please emit an error, that _could_ be displayed somewhere, in a dialog box or so.

aldipower replied:
You are catching there lines above. But that is not good practice.
Somebody other could unaware of this beware of the function and don't catch it accidentally.

orlandohohmeier replied:
How do I fail those methods in a way that helps with debugging but didn't break the app?

aldipower replied:
Please don't let it fail.

aldipower replied:
The app should handle an undefined state gracefully.

orlandohohmeier replied:
This way I need to add a bunch of extra value checks, and I never get any meaningful response why the value is undefined.

aldipower replied:
Yes, please add these checks. This makes the code more descriptive, too.

orlandohohmeier replied:
Hm. I disagree, having a null/undefined check doesn't tell why it's undefined... I would like to have this information somehow available. Anyway, I'll rewrite the error conditions using null/undefined values.

aldipower replied:
Thx.

aldipower replied:
Note: JavaScript is not Java. ;)
Exception handling isn't solved that good in JavaScript, it lacks of some mandatory features.

aldipower replied:
That isn't possible in JavaScript:
http://www.javatpoint.com/throws-keyword-and-difference-between-throw-and-throws

orlandohohmeier replied:
But simply removing the valuable information, because exception handling isn't as good as it should be, doesn't sound like a good practice either. Anyway, let's discuss at other times.  :smile:

aldipower replied:
No, let's discuss it here, it seems there is a demand for discussion.
And I think you didn't got my point.
So, which valuable information do you mean? The user doesn't see the exception message at all. He only sees the crashed application, if you forgot to handle the exception correctly.
At the other hand, you do handling the exception always! So the message is completely meaningless, it will never be printed out, you just discard it.

orlandohohmeier replied:
Those exception messages were meant for developers, hopefully helping in case of failure. Without those exceptions I have to read through whole `getNodeUrl` method adding a bunch of breakpoints or logs to actually know why the result is undefined. Idealy we should also have a meaningful display for the user.

aldipower replied:
OK, I see your point.
Delivering debug code in production isn't an option either. :)

orlandohohmeier replied:
I'm going to refactor this, in order to get this PR merged and the feature shipped in time, but would like to discuss this at some point as I see value in having exceptions. BTW: The libs and frameworks we use have a lot of exception. :)

aldipower replied:
OK, then let's discuss this at another point.
I am not against exceptions in general, I think we use them at some places, where the application should break, e.g.
And a lot of libraries we are using aren't that good, I looked into qajax, e.g., that code is really ####.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  });
+
+  if (agent == null) {
+    throw new Error(
```
This breaks the application.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  var agentState = MesosStore.getState(agentId);
+
+  if (agentState == null) {
+    throw new Error(
```
No way. Please emit an error.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    // Get node/agent sate. This is needed to actually retrieve
+    // the executor directory.
+    if (!MesosStore.getState(agentId)) {
+      try {
```
Please get rid of the try-catch. Thx.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    agentState.completed_frameworks.find(matchFramework);
+
+  if (!framework) {
+    throw new Error(
```
Same here.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    framework.completed_executors.find(matchExecutor);
+
+  if (executor == null) {
+    throw new Error(
```
Same here.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+
+    // Request the files if not present
+    if (!MesosStore.getTaskFiles(taskId)) {
+      try {
```
Get rid of it.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+
+var MesosStore = Object.assign({
+
+  getState: function (nodeId) {
```
Maybe ```getMesosState``` ?

orlandohohmeier replied:
It's `MesosStore.getState` isn't that explicit enough?

aldipower replied:
Yes, it is. :)

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+var MesosActions = require("../actions/MesosActions");
+var MesosEvents = require("../events/MesosEvents");
+
+const FILES_TTL = 60000;
```
Yeah, a meaningful usage of ```const``` !! :)

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  var data = action.data;
+  switch (action.actionType) {
+    case MesosEvents.REQUEST_FILES:
+      taskFileRequestQue.push({
```
```queue``` instead of ```que```

pierlo-upitup commented on package.json:
```diff
@@ -23,6 +23,7 @@
     "moment": "^2.9.0",
     "mousetrap": "^1.4.6",
     "object-path": "^0.9.2",
+    "q": "^1.4.1",
```
We need to introduce the `q` dependency in our shrinkwrap too. Could you rebase this PR against master, do an `rm -rf node_modules && npm install` before doing a `npm run shrinkwrap` and add the resulting changes to the `npm-shrinkwrap.json`? 
cc @aldipower for a doublecheck of what I just wrote

aldipower replied:
You got rid of this, didn't you?

pierlo-upitup replied:
Oh yeah. Simply remove the line then!

aldipower commented on src/js/components/TaskDetailComponent.jsx:
```diff
@@ -186,6 +187,9 @@ var TaskDetailComponent = React.createClass({
         {this.getTaskHealthComponent()}
+        <hr />
+        <h3>Working Directory</h3>
+        <TaskFileListComponent task={task}/>
```
Extra space behind } !

Check your editor warnings.

orlandohohmeier replied:
I don't got any eslint warning for this... :crying_cat_face:

aldipower commented on src/js/components/TaskDetailComponent.jsx:
```diff
@@ -186,6 +187,9 @@ var TaskDetailComponent = React.createClass({
           <dd><TaskMesosUrlComponent task={task}/></dd>
```
Oh, this also needs an extra spaces.

aldipower commented on src/css/components/task-list.less:
```diff
@@ -4,7 +4,7 @@
 .table.task-list {
   .healthy {
-    color:@task-health-healthy-color;
+    color: @task-health-healthy-color;
```
Thank you! :)

aldipower commented on src/css/components/task-file-download.less:
```diff
@@ -0,0 +1,9 @@
+ */
+.task-file-download {
+  &.loading  {
+    // @todo: add some styles indicating that somethig is going on
```
I think this will happen in this PR?
Otherwise, no unused code.
Remove it.

pierlo-upitup replied:
Agreed. I suggest to move the loading states to a separate PR and concentrating on integrating comments on this one ;)

orlandohohmeier replied:
Ok then, let's do the loading state in separate PR. I'll remove this file.

pierlo-upitup commented:
Please don't forget to add the CHANGELOG entry.

pierlo-upitup commented:
I would say that the loading state indicators could belong to another PR if that's fine by you, @orlandohohmeier. We should create a specific issue for it, of course!

aldipower commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    var taskId = task.id || task.taskId;
+    var files = MesosStore.getTaskFiles(taskId);
+    if (files) {
+      return files.filter(matchFileName(props.fileName))[0];
```
matchFileName is only use once, please get rid of it.

aldipower commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+  getInitialState: function () {
+    return {
+      file: this.getFile(),
+      requested: false
```
Please be more precise. Request what?

pierlo-upitup replied:
Also, I personally really like using `is` / `has` prefixes for booleans where appropriate.

aldipower commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+
+  getFile: function () {
+    var props = this.props;
+    var task = this.props.task;
```
Use ```props``` not ```this.props```

aldipower commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    }
+  },
+
+  render: function () {
```
Cache this.state and this.props into variables, it's both used twice.

pierlo-upitup commented on src/css/components/task-file-list.less:
```diff
@@ -0,0 +1,22 @@
+ * ==============
+ */
+.task-file-list {
+  margin-top: @base-spacing-unit * 3;
```
remove spaces, we decided to keep it tight: `@base-spacing-unit*3`

pierlo-upitup commented on src/css/components/task-file-list.less:
```diff
@@ -0,0 +1,22 @@
+  margin-top: @base-spacing-unit * 3;
+
+  tr {
+    .btn {
```
Would it make sense to create a `.btn.reveal` class to handle the "show on hover" case? I find the nesting quite inexpressive in this case.

orlandohohmeier replied:
Do you mean an extra class that's added/removed by react?

pierlo-upitup replied:
No, I simply mean to have a reusable class `.reveal` that we can apply on `.btn` that makes the button appear on hover like you're doing here, instead of having to specify this behavior in a one-off way as u've done here.

pierlo-upitup commented on src/css/components/task-file-list.less:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    &:hover {
+      .btn {
```
also why the `.btn .btn` nesting?

pierlo-upitup replied:
nvm, I read this wrong.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    state: state,
+    timestamp: Date.now()
+  };
+  MesosStore.emit(MesosEvents.CHANGE);
```
Does it makes sense to split the CHANGE event into two?
This here could be ```MesosEvents.CHANGE_STATE``` and below it could be ```MesosEvents.CHANGE_TASKFILES``` ?

Then you could probably write the components even cleaner.

orlandohohmeier replied:
Yeah, I was actually thinking about that, but I wasn't sure how this fits into the flux architecture.

aldipower replied:
Where should be the difference in the architecture?

orlandohohmeier replied:
We're losing the single change event which is fired for every single change... Anyway, I'll refactor this, as two different events feels way better.

aldipower replied:
:+1:

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    files: files,
+    timestamp: Date.now()
+  };
+  MesosStore.emit(MesosEvents.CHANGE);
```
```MesosEvents.CHANGE_TASKFILES``` ?

pierlo-upitup commented on src/css/main.less:
```diff
@@ -31,4 +31,7 @@
 @import "components/pane.less";
 @import "components/panel.less";
+@import "components/task-list.less";
+@import "components/task-file-download.less";
```
Remove if we decide to extract the loading states to another PR

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  });
+}
+
+AppDispatcher.register(function (action) {
```
In generally we try to emit all events from this place.

This means the private store functions are encapsulated, stands for themselves and don't emit directly from the store. They return values if necessary. So at the in the ```case``` you could emit. That holds the things much more at one place and is better readable.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+var taskFileMap = {};
+var taskFileRequestQue = [];
+
+var MesosStore = Object.assign({
```
Could you move this below the private functions. Thx.
Otherwise it will generate warnings, that the functions are not defined yet.

aldipower commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    }
+    return (
+      <a className={className}
+        href={href}
```
+2 itendation

aldipower commented on src/js/components/TaskListComponent.jsx:
```diff
@@ -178,6 +178,12 @@ var TaskListComponent = React.createClass({
                 </span>
               </th>
+              <th className="text-center">
+                  Error Log
```
-2 idendation

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  }
+
+  var executor =
+    framework.executors.find(matchExecutor) ||
```
This fits into the first line, raise it up.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+      break;
+    case MesosEvents.REQUEST_FILES_COMPLETE:
+      addTaskFiles(data.id, data.files.map((file) => {
+        let encodedPath = encodeURIComponent(file.path);
```
```var``` instead of ```let``` here.

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    var task = this.props.task;
+    var taskId = task.id || task.taskId;
+    var files = MesosStore.getTaskFiles(taskId);
+    if (files) {
```
How about a more explicit `if (files != null && files.length)` ?

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    var className = classNames("task-file-download", this.props.className, {
+      "loading": this.state.requested
+    });
+    var file = this.state.file;
```
please cache `this.state`

pierlo-upitup replied:
and this.props too

aldipower commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+      <table className="table table-hover table-selectable task-file-list">
+        <thead>
+        <th>
+          <span onClick={this.sortBy.bind(null, "name")}
```
We indent this way:
```html
<span onClick={this.sortBy.bind(null, "name")}
    className={headerClassSet}>
  ID {this.getCaret("name")}
</span>
```

orlandohohmeier replied:
@aldipower @pierlo-upitup Do you know how to configure Intellij to do this indention automatically?

pierlo-upitup replied:
Not sure. Maybe @philipnrmn ?

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+
+var TaskFileLinkComponent = React.createClass({
+  displayName: "TaskFileLinkComponent",
+  propTypes: {
```
add a trailing empty line here pls!

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+  return (file) => file.name === name;
+}
+
+var TaskFileLinkComponent = React.createClass({
```
Shouldn't this component be called like the filename, ie `TaskFileDownloadComponent` ??

orlandohohmeier replied:
Ups... Forgot to adjust the component during refactoring.

pierlo-upitup commented on src/js/components/TaskListItemComponent.jsx:
```diff
@@ -193,6 +195,16 @@ var TaskListItemComponent = React.createClass({
+        </td>
+        <td className="text-center">
+          <TaskFileDownloadComponent task={task} fileName="stdout">
+            <i className="icon icon-mini file" /> stdout
```
Given that the file download text reads exactly like the `fileName` prop, and that the icon represents a file, why is this component responsible for the download icon and text nodes? Couldn't we simplify and move this into the TaskFileDownloadComponent itself? What do you think @orlandohohmeier ?

orlandohohmeier replied:
Yeah, makes sense.

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    // Start download if file was requested by the user
+    if (file && requested) {
+      window.open(file.download);
+      requested = false;
```
You are mutating state directly this way. Please simply remove this line, and return `requested: false` in the `setState` call.

orlandohohmeier replied:
As the change event is fired quite often, we can't update the request state unless we have the correct file, thus we can't simply remove the line. Besides that, it's a primitive, thus it's copied and the state isn't mutated.

pierlo-upitup replied:
You're right! Pardon me!

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+
+  render: function () {
+    var className = classNames("task-file-download", this.props.className, {
+      "loading": this.state.requested
```
please cache this.state and this.props whenever possible.

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+    if (file == null) {
+      var task = this.props.task;
+      var agentId = task.slaveId;
+      var taskId = task.id || task.taskId;
```
these should all be `let`s.

pierlo-upitup commented on src/js/components/TaskFileDownloadComponent.jsx:
```diff
@@ -0,0 +1,98 @@
+      this.setState({
+        requested: true
+      });
+      event.preventDefault();
```
I'd prefer this to be moved to the beginning of the `if` block, i think it feels more...natural?

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+
+var TaskFileListComponent = React.createClass({
+  displayName: "TaskFileListComponent",
+  propTypes: {
```
always add a white line between these

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+
+  getInitialState: function () {
+    var task = this.props.task;
+    var taskId = task.id || task.taskId;
```
Having to perform this check on each task really annoys me. Can we think of a better solution? How about "preprocessing" tasks to make sure we can always count on task.id ? @aldipower what do you think?

aldipower replied:
Good idea, should be handled in the store.
Actually I wonder why this check is needed at all?
Why this isn't consistent?

orlandohohmeier replied:
I only added those checks is they were implemented within the [`TaskMesosUrlComponent`](https://github.com/mesosphere/marathon-ui/blob/master/src/js/components/TaskMesosUrlComponent.jsx#L55) and I wanted to be safe... but yeah, let's move that check to the `AppsStore``

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+    };
+  },
+
+  componentWillMount: function () {
```
[super nitpick mode on]
`componentDidMount` should come first
[super nitpick mode off]

orlandohohmeier replied:
I thought we order them by the lifecycle... :confused:

1. componentWillMount 
2. (render)
3. componentDidMount

pierlo-upitup replied:
That sounds new to me? `render` is always last...

orlandohohmeier replied:
Yeap, but I thought the render method is the only exception and the rest is ordered by the lifecycle. I tried to illustrate that with list, but obviously failed. :smile:

pierlo-upitup replied:
Oh I see what you mean. OK I think we never really had a decision there, so I guess lifecycle order makes sense. I just like to see the willMount/willUnmount close when they do the bind/unbind dance. Feel free to ignore my comment ;)

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+    if (files != null) {
+      return lazy(files)
+        .sortBy((app) => {
+          return app[sortKey];
```
`.sortBy(app => app[sortyKey]`

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+        .map((file) => {
+          let lastModifiedDate = new Date(file.mtime);
+          let lastModifiedIsoString = lastModifiedDate.toISOString();
+          let lastModifiedLocaleString = lastModifiedDate.toLocaleString();
```
as this is function-scope they should probably be `var`s

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+              </td>
+              <td className="text-right">
+                <span>
+                  {`${Util.filesize(file.size * Math.pow(1024, 2), 0)}`}
```
:+1:

pierlo-upitup commented on src/js/components/TaskFileListComponent.jsx:
```diff
@@ -0,0 +1,179 @@
+        <th>
+          <span onClick={this.sortBy.bind(null, "name")}
+            className={headerClassSet}>
+                ID {this.getCaret("name")}
```
Mind the indentation here and following spans...should be something like

```
<span onClick={this.sortBy.bind(null, "name")}
    className={headerClassSet}>
  ID {this.getCaret("name")}
</span>
```

pierlo-upitup replied:
sorry, double comment. had to reload the page to see the comment before mine...

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+var MesosEvents = require("../events/MesosEvents");
+
+const FILES_TTL = 60000;
+const STATE_TTL = 60000;
```
We should probably ask for feedback on these values at some point.

orlandohohmeier replied:
Yep. We probably could cache the data even longer (where applicable) as we do invalidation not only on time but also if data is missing.

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+
+}, EventEmitter.prototype);
+
+function addState(nodeId, state) {
```
`setStateForNodeId`?

pierlo-upitup replied:
or even `updateState` ?

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  // Get master sate, We need it to parse the agent/node url
+  if (!MesosStore.getState(MASTER_ID)) {
+    MesosActions.requestState(MASTER_ID,
+      info.marathon_config.mesos_leader_ui_url.replace(/\/?$/, "/master"));
```
What if for some reason mesos_leader_ui_url is undefined or empty, btw?

orlandohohmeier replied:
Overlooked that one. May bad.

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+    return;
+  }
+
+  // Check all requests, request needed data or remove them from the que
```
queue

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+  }
+
+  // Check all requests, request needed data or remove them from the que
+  taskFileRequestQue.forEach((request, index) => {
```
Queue

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+
+  // Check all requests, request needed data or remove them from the que
+  taskFileRequestQue.forEach((request, index) => {
+
```
remove newline

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+function resolveTaskFileRequests() {
+  var info = InfoStore.info;
+
+  // Get the marathon config, to retrieve the mesos leader (master)  url
```
Could you remove these comments. They should not be needed.
The code should be self-explanatory, otherwise the code should be improved/refactored.

pierlo-upitup commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+      MesosStore.emit(MesosEvents.REQUEST_STATE_ERROR, data.body);
+      break;
+    case MesosEvents.REQUEST_FILES_COMPLETE:
+      addTaskFiles(data.id, data.files.map((file) => {
```
btw with single args you can skip the parenthesi

pierlo-upitup commented:
Great stuff, @orlandohohmeier ! Can't wait to see this merged. It will be probably one of the most impactful features we've released in a while for a big chunk of our users!

aldipower commented:
Also a rebase from master would be neat to have the new npm and node in.

aldipower commented:
Due to my configuration I get a 404 on http://dcos:5050/master/state .
That's ok, but I need an alert box to be informed on that.
On a 404 the user click on the link does nothing. 
That isn't ideal in terms of response to the user.

aldipower commented on src/js/actions/MesosActions.js:
```diff
@@ -0,0 +1,48 @@
+var MesosEvents = require("../events/MesosEvents");
+
+var MesosActions = {
+  requestState: function (id, url) {
```
Which ```id``` is this? taskId? Really don't know it.
Could you be more precise. Thx.

orlandohohmeier replied:
The id is only passed along to identify the request, you can use whatever you want to reference the it.

aldipower replied:
Ah ok! Cool.
Then maybe name it ```requestId``` to make it clear?

orlandohohmeier commented:
Good point! We need an error state for the list and download buttons.

aldipower commented on src/js/actions/MesosActions.js:
```diff
@@ -0,0 +1,48 @@
+
+var MesosActions = {
+  requestState: function (id, url) {
+    JSONPUtil.request(`${url}/state`).then(
```
OK ```/state/``` doesn't work with Mesos 0.24, it must be ```/state.json```. Then it works!
Could you test if state.json is also working with 0.25? Then everything is perfect.

orlandohohmeier replied:
The `state.json` endpoint works, but it's already [marked as deprecated](https://github.com/apache/mesos/blob/ffbed5ea3bb059ed6dd8830d8a6acc5195ed3683/src/master/master.cpp#L826) and will be removed in 0.26.

aldipower replied:
Ah ok, then you need to detect the mesos version somehow, to be compatible to all these versions.

aldipower commented on src/js/actions/MesosActions.js:
```diff
@@ -0,0 +1,48 @@
+  },
+  requestFiles: function (id, host, filePath) {
+    JSONPUtil.request(
+      `${host}/files/browse?path=${encodeURIComponent(filePath)}`)
```
OK ```/browse``` doesn't work with Mesos 0.24, it must be ```/browse.json```. Then it works!
Could you test if browse.json is also working with 0.25. Then everything is perfect.

aldipower commented on src/js/stores/MesosStore.js:
```diff
@@ -0,0 +1,221 @@
+        let encodedPath = encodeURIComponent(file.path);
+        file.host = data.host;
+        file.name = /[^/]+\/?$/.exec(file.path)[0];
+        file.download = `${data.host}/files/download?path=${encodedPath}`;
```
```/download``` doesn't work with Mesos 0.24, ```/download.json``` does.

aldipower replied:
Please rename ```file.download``` to ```file.downloadURI``` .

pierlo-upitup commented:
Thanks for all the updates, @orlandohohmeier ! What happened to `npm-shrinkwrap.json`, though? Ideally we should only see the lines relevant to `semver` being added... Did you rebase master into this PR?

orlandohohmeier commented:
@pierlo-upitup Yeah, I rebased master into this branch and updated everything accordingly. To be honest, I don't know why shrinkwrap added all the `from` entries. But as they seem to reflect the `resolved` entries I assume it shouldn't do any harm.

pierlo-upitup commented:
@orlandohohmeier OK, GitHub can't handle the size of it so I assume if it's just the `from` fields being added that's fine. Still, I wonder why those weren't added in https://github.com/mesosphere/marathon-ui/pull/462 ...

aldipower commented:
This PR takes a while to load now.. ^^

orlandohohmeier commented:
@aldipower Yep :joy_cat:

bhenderson commented:
sorry to comment on an already large PR, but any reason why you are downloading the files, instead of showing them in the browser a la the mesos ui?

orlandohohmeier commented:
@bhenderson No worries. We started with the file download, as it is the minimum viable implementation. Reading the logs à la Mesos UI will be part of the next iteration.

bhenderson commented:
sounds good, thanks @orlandohohmeier for the explanation.

orlandohohmeier commented:
It takes a while to load and read through all the comments. Thus I'm closing this PR and open a new one as discussed.
